### PR TITLE
[codex] Fix stale wait replay race

### PR DESCRIPTION
**Pull request by pranaygp**
## Summary

Fixes the stale wait-triggered replay race for `Promise.race([hook, sleep])` without doing a full event-log reload on the common path.

When the workflow handler loads events, then a hook payload lands durably before the handler commits `wait_completed`, replay must use the updated durable event order. This now uses the cursor at the end of the original event snapshot to fetch only events created after that snapshot, so `hook_received` before `wait_completed` is observed before replay.

## Changes

- Removes the prior seeded `hook-sleep-interaction.test.ts` repro that fed `runWorkflow()` an already-corrupt event log.
- Removes the timing/seeded e2e repro from `99_e2e.ts` and `e2e.test.ts` for now.
- Adds a deterministic handler-level unit test around `workflowEntrypoint()` that simulates the real interleaving at the World boundary.
- Adds optional `cursor` and `hasMore` pagination metadata to `EventResult` for `run_started` preloaded events, matching the regular event listing response shape.
- Local/postgres worlds populate the pagination metadata; world-vercel accepts and forwards it when the backend provides it.
- Refreshes workflow events after elapsed wait completion by fetching only events after the original snapshot cursor, with full-refetch fallback when the cursor is missing or the delta does not contain the attempted wait completions.
- Adds a patch changeset for the changed packages.

## Validation

- Confirmed the handler repro failed before the original runtime fix: `expected listEvents to be called 2 times, but got 1 times`.
- `pnpm vitest run packages/core/src/runtime/wait-completion-replay.test.ts`
- `pnpm vitest run packages/core/src/hook-sleep-interaction.test.ts`
- `pnpm --filter @workflow/world build`
- `pnpm --filter @workflow/core typecheck`
- `pnpm --filter @workflow/world-local typecheck`
- `pnpm --filter @workflow/world-postgres typecheck`
- `pnpm --filter @workflow/world-vercel typecheck`
- `pnpm changeset status --since=stable`
- `pnpm biome check --files-ignore-unknown=true packages/core/src/runtime.ts packages/core/src/runtime/helpers.ts packages/core/src/runtime/wait-completion-replay.test.ts packages/world/src/events.ts packages/world-local/src/storage/events-storage.ts packages/world-postgres/src/storage.ts packages/world-vercel/src/events.ts .changeset/bright-waits-refresh.md` (passes with existing warnings in large runtime/storage files)

### CODE REVIEW

File: packages/core/e2e/e2e.test.ts [hunk 1]
```diff
@@ -17,7 +17,7 @@ import {
   test,
 } from 'vitest';
 import { getTrustedSourcesHeaders } from '../../../scripts/trusted-sources-headers.mjs';
-import type { Run } from '../src/runtime';
+import type { Run as WorkflowRunHandle } from '../src/runtime';
```
Comment (pranaygp) [inline packages/core/e2e/e2e.test.ts]: this change is now redundant

File: .changeset/bright-waits-refresh.md [hunk 2]
```diff
@@ -0,0 +1,6 @@
+---
+'@workflow/core': patch
+'workflow': patch
```
Comment (pranaygp) [inline .changeset/bright-waits-refresh.md]: remove `workflow` and just keep `@workflow/core`

File: packages/core/e2e/e2e.test.ts (hunk 1 above)
Comment (pranaygp) [reply packages/core/e2e/e2e.test.ts]: Fixed in 2ed07e90d: reverted this alias churn, so `packages/core/e2e/e2e.test.ts` now has no net diff against `stable`.

File: packages/core/src/runtime/wait-completion-replay.test.ts [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+  const runId = 'wrun_stale_wait_replay';
+  const workflowName = 'workflow';
+  const deploymentId = 'dpl_stale_wait_replay';
+  const hookToken = 'stale-wait-hook-token';
+  const startedAt = new Date('2026-05-19T12:00:00.000Z');
+  const workflowArgs = await dehydrateWorkflowArguments(
+    [hookToken],
+    runId,
+    undefined
+  );
```
[the remark this answers is not in the document]
Comment (pranaygp) [reply packages/core/src/runtime/wait-completion-replay.test.ts:55]: Fixed in 2ed07e90d: the test now mocks Date.now() to a fixed time before the elapsed wait check, so waitsToComplete is deterministic regardless of the machine clock.

File: .changeset/bright-waits-refresh.md (hunk 2 above)
Comment (pranaygp) [reply .changeset/bright-waits-refresh.md]: Fixed in 2ed07e90d: removed the direct `workflow` entry from this changeset and kept the package-specific entries for the packages this PR changes.

File: packages/core/e2e/e2e.test.ts (hunk 1 above)
Comment (pranaygp) [reply packages/core/e2e/e2e.test.ts]: Fixed in 2ed07e90d: reverted the remaining e2e alias churn; this file now has no net diff against stable.

**Comment by karthikscale3** [comment]
AI Review:

## The Bug

**Scenario:** A workflow does `Promise.race([hook, sleep(5s)])`. The race condition unfolds like this:

1. Wait timer fires → handler boots → calls `run_started` which returns a snapshot of events (cursor=C1). Snapshot does **not** yet include `hook_received`.
2. Between snapshot load and the elapsed-wait scan, the hook payload lands durably (`hook_received` row commits to the DB).
3. Handler checks elapsed waits, sees the wait passed `resumeAt`, creates `wait_completed`. The pre-fix code just appended this in-memory to the stale events array.
4. Replay sees: `wait_created`, then `wait_completed` (no `hook_received`) → sleep branch wins.

But the **durable order** in the DB is `hook_received` → `wait_completed`, so on the *next* cold replay the workflow would observe the hook branch winning. **Non-determinism — replay diverges from durable history.**

## The Fix

The runtime now treats the `run_started`-preloaded events as a **snapshot with a cursor** rather than as the complete truth:

1. `EventResult` gains an optional `eventsCursor` (worlds populate it; postgres uses the last `eventId`, local uses its paginator's cursor).
2. After creating `wait_completed`, instead of just pushing the new event to memory, the runtime **re-queries events after the snapshot cursor**.
3. A guard checks that the delta actually contains the `wait_completed` events it just made (`sawAllWaitCompletions`). If yes, splice the delta in. If no (stale cursor, replica lag, missing world support), **fall back to a full reload**.
4. If the world didn't supply a cursor at all, also fall back to a full reload.

This gives the runtime the *durable* order of `hook_received` vs `wait_completed`, so the replay matches what a future cold replay would see.

## Is the fix correct?

**Yes, the approach is sound.** The cursor model is the right abstraction, the consistency check (`sawAllWaitCompletions`) is a smart safety net, and the fallback paths preserve correctness even on older worlds. Tests cover all three branches (cursor present + works, no cursor, cursor but stale delta).

## Other scenarios worth thinking through

A few I'd want to confirm:

1. **Postgres cursor uses `eventId` (ULID), not commit order.** In `packages/world-postgres/src/storage.ts:1326`, pagination is `WHERE eventId > cursor` with `ORDER BY eventId`. If a `hook_received` was *generated* (ULID minted) before the snapshot's last event but *committed* after (slow insert / cross-process delay), its ULID could be lexicographically smaller than the cursor — the delta query would silently skip it. The `sawAllWaitCompletions` check wouldn't catch this because our `wait_completed` (always newer ULID) would still be present. The full-reload fallback would also miss it because the full reload paginates by the same `eventId` field. This isn't introduced by the PR (existing pagination has the same property), but the fix doesn't fully close the door on it either. Worth verifying how serialized hook delivery is in practice.

2. **Read-replica lag on the cursor query.** If the postgres backend serves the delta read from a follower that hasn't replicated the `wait_completed` insert yet, `sawAllWaitCompletions` is false → full reload. Good — but the full reload likely hits the same follower. May be fine because `events.create` likely targets primary and subsequent reads in the same handler stick to primary, but worth confirming.

3. **Multiple elapsed waits + interleaved durable events.** If two waits elapse in the same tick and a `hook_received` lands between the two `wait_completed` creates, durable order is `wc_A, hook_received, wc_B`. The delta query returns all three in order, replay is correct. ✓ (the test only exercises one wait — adding a multi-wait variant would be nice.)

4. **The race isn't unique to hooks.** Any externally-written event (signal events, future event types) interleaved between snapshot and wait completion has the same race. The fix is generic (re-reads everything after cursor) so it covers them, which is good.

5. **Lost handler attempt scenario.** Handler A loads snapshot, gets preempted; Handler B completes everything; Handler A resumes, hits `EntityConflictError` on every `wait_completed` create. The code still enters the `waitsToComplete.length > 0` branch and re-queries — `sawAllWaitCompletions` will be true (B created them), so A picks up the full state via the delta. Correct, though A's replay is wasted work.

6. **Cursor preserved across paginated snapshots.** `getAllWorkflowRunEvents` returns the cursor from the *last* page. ✓. But `events?: Event[]` from `run_started` is only populated when results fit a single page (postgres impl reads with no limit, local uses its paginator). If a future change paginates the preloaded events but only returns page 1's cursor while including all pages' events, the cursor would point inside the snapshot — `sawAllWaitCompletions` would be true but the delta would double-count events already in `events`. Currently safe, but a footgun for future edits.

7. **Minor:** `eventsCursor = newEvents.cursor ?? eventsCursor` in `runtime.ts` is a bit unusual — if the delta query returned 0 events with a `null` cursor, we'd keep the old cursor for any future use. Today there is no "future use" within the same handler invocation, so it's harmless, but a comment would help.

Overall: ship it, but I'd at least add a multi-wait + interleaved-hook test, and double-check the ULID-cursor / late-commit edge case is genuinely unreachable in production (or document why).

File: packages/core/src/runtime.ts [hunk 4]
```diff
@@ -440,6 +453,47 @@ export function workflowEntrypoint(
...
+                        .map((e) => e.correlationId)
+                    );
+                    const sawAllWaitCompletions = waitsToComplete.every(
+                      (waitEvent) =>
+                        completedWaitIdsAfterCursor.has(waitEvent.correlationId)
+                    );
+
+                    if (sawAllWaitCompletions) {
+                      events.push(...newEvents.events);
+                      eventsCursor = newEvents.cursor ?? eventsCursor;
```
Comment (VaguelySerious) [inline packages/core/src/runtime.ts]: AI Review: Nit

`eventsCursor` is written here and again on lines 486/493, but never read after this block — the three assignments are dead. Also, `newEvents.cursor ?? eventsCursor` is unreachable as written: `sawAllWaitCompletions` requires the delta to contain at least one `wait_completed` event, so `newEvents.cursor` is guaranteed non-null when this branch runs.

Safe to drop `eventsCursor` from these branches entirely.

File: packages/world-local/src/storage/events-storage.ts [hunk 5]
```diff
@@ -962,6 +964,8 @@ export function createEventsStorage(
           getId: (e) => e.eventId,
         });
         events = allEvents.data;
+        cursor = allEvents.cursor;
+        hasMore = allEvents.hasMore;
```
Comment (VaguelySerious) [inline packages/world-local/src/storage/events-storage.ts:969]: AI Review: Nit

`paginatedFileSystemQuery` defaults `limit` to 20 and this preload call doesn't override it. In practice the first `run_started` only has 1-2 events so this never trips, and the new `hasMore` gate in the runtime correctly falls back to a full reload if it ever did — but the intent at this call site is "give me everything", so worth passing an explicit large limit to match.

File: packages/core/src/runtime/wait-completion-replay.test.ts [hunk 6]
```diff
@@ -0,0 +1,429 @@
...
+      'step_created',
+      'step_started',
+      'step_completed',
+      'wait_created',
+      'hook_received',
+      'wait_completed',
+    ]);
+    expectHookBranchQueued(result);
+  });
+});
```
Comment (VaguelySerious) [inline packages/core/src/runtime/wait-completion-replay.test.ts:473]: AI Review: Nit

Missing a fourth case for `hasMore=true`: when the preload response includes events but flags more pages exist, the runtime should discard the partial preload and do a full `events.list` with `cursor: undefined`. I validated this works locally — the gate at `runtime.ts:309-316` is the one branch not exercised by the existing three cases.

**Review by VaguelySerious — APPROVED** [review]
AI review: no blocking issues

File: packages/world-local/src/storage/events-storage.ts (hunk 5 above)
Comment (pranaygp) [reply packages/world-local/src/storage/events-storage.ts:969]: Fixed in 5b9ee65db: the local run_started preload query now passes an explicit limit: 1000 so the call site matches its intent of returning a complete snapshot when possible.

File: packages/core/src/runtime/wait-completion-replay.test.ts (hunk 6 above)
Comment (pranaygp) [reply packages/core/src/runtime/wait-completion-replay.test.ts:473]: Fixed in 5b9ee65db: added a fourth handler-level case where run_started returns preloaded events with hasMore: true. The test asserts the runtime discards that partial preload, does a full events.list from the beginning, then uses the resulting cursor for the post-wait delta fetch.

File: packages/core/src/runtime.ts (hunk 4 above)
Comment (pranaygp) [reply packages/core/src/runtime.ts]: Fixed in 5b9ee65db: removed the dead eventsCursor reassignments after the post-wait refresh. The cursor is only needed to choose the refresh strategy before replay.
